fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingHeartbeat.java

```diff
@@ -0,0 +1,186 @@
+/*
+ * IndexingHeartbeat.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2025 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.provider.foundationdb;
+
+import com.apple.foundationdb.KeyValue;
+import com.apple.foundationdb.async.AsyncIterator;
+import com.apple.foundationdb.async.AsyncUtil;
+import com.apple.foundationdb.record.IndexBuildProto;
+import com.apple.foundationdb.record.logging.LogMessageKeys;
+import com.apple.foundationdb.record.metadata.Index;
+import com.apple.foundationdb.synchronizedsession.SynchronizedSessionLockedException;
+import com.google.protobuf.InvalidProtocolBufferException;
+
+import javax.annotation.Nonnull;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class IndexingHeartbeat {
+    // [prefix, xid] -> [indexing-type, genesis time, heartbeat time]
+    final UUID indexerId;
+    final IndexBuildProto.IndexBuildIndexingStamp.Method indexingMethod;
+    final long genesisTimeMilliseconds;
+    final long leaseLength;
+
+    public IndexingHeartbeat(final UUID indexerId, IndexBuildProto.IndexBuildIndexingStamp.Method indexingMethod, long leaseLength) {
+        this.indexerId = indexerId;
+        this.indexingMethod = indexingMethod;
+        this.leaseLength = leaseLength;
+        this.genesisTimeMilliseconds = nowMilliseconds();
+    }
+
+    public void updateHeartbeat(@Nonnull FDBRecordStore store, @Nonnull Index index) {
+        byte[] key = IndexingSubspaces.indexheartbeatSubspace(store, index, indexerId).pack();
+        byte[] value = IndexBuildProto.IndexBuildHeartbeat.newBuilder()
+                .setMethod(indexingMethod)
+                .setGenesisTimeMilliseconds(genesisTimeMilliseconds)
+                .setHeartbeatTimeMilliseconds(nowMilliseconds())
+                .build().toByteArray();
+        store.ensureContextActive().set(key, value);
+    }
+
+    public CompletableFuture<Void> checkAndUpdateHeartbeat(@Nonnull FDBRecordStore store, @Nonnull Index index) {
+        // complete exceptionally if non-mutual, other exists
+        switch (indexingMethod) {
+            case SCRUB_REPAIR:
+            case MUTUAL_BY_RECORDS:
```

Comment:
Should `MUTUAL_BY_RECORDS` really be unlimited?
Are you intending to have a more interesting policy in the future?
Particularly, if there are 10 boundary keys, it doesn't make sense to have more than 11 concurrent indexers.
This plays with my comment above, if this takes a max-concurrency, one could imagine that
- `IndexingByIndex` and `IndexingMultiTargetByRecords` pass in 1
- `IndexScrubbing` passes in one value if it is repairing, a different value if it is not, perhaps one of those could be `Integer.MAX_VALUE`
- `IndexingMutuallyByRecords` passes in the number of fragments (maybe +1)